pkg/apis/ext.cattle.io/v1/types.go

```diff
@@ -26,3 +27,148 @@ type UserActivityStatus struct {
 	// +optional
 	ExpiresAt string `json:"expiresAt"`
 }
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Token is used to authenticate requests to Rancher.
+type Token struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the Token.
+	Spec TokenSpec `json:"spec"`
+	// Status is the most recently observed status of the Token.
+	Status TokenStatus `json:"status"`
+}
+
+// TokenSpec defines the desired state of the Token.
+type TokenSpec struct {
+	// UserID is the kube resource id of the user owning the token. By
+	// default that is the user who owned the token making the request
+	// creating this token. Currently this default is enforced, i.e. using a
+	// different user is rejected as forbidden.
+	// +optional
+	UserID string `json:"userID,omitempty"`
+	// UserPrincipal holds the information about the ext auth provider
+	// managed user (principal) owning the token.
+	UserPrincipal TokenPrincipal `json:"userPrincipal""`
```

Comment:
not sure this is an error, but the json field tag has two `"` at the end: `json:"userPrincipal""`
was the `omitempty` intentionally missing in this case?

Reply:
Fixed now, not yet pushed. The structure should never be empty (`Name` ahould be always set), so an `omit...` does not make sense.

Reply:
pushed e360a27a6bd379040f3f4e1193ddf28f49690377